--- a/src/jpeg.inl
+++ b/src/jpeg.inl
@@ -923,7 +923,7 @@
                 while (iLen)
                 {
                     //               if (iBitNum > 6) // do long table
-                    if ((cc >> (iBitNum-5)) == 0x1f) // first 5 bits are 1 - use long table
+                    if (iBitNum >= 5 && (cc >> (iBitNum-5)) == 0x1f) // first 5 bits are 1 - use long table
                     {
                         count = iMaxLength - iBitNum;
                         codestart = cc << count;
@@ -1011,7 +1011,7 @@
                 iLen = *pBits++; // get number of codes for this bit length
                 while (iLen)
                 {
-                    if ((cc >> (iBitNum-4)) == 0xf) // first 4 bits are 1 - use long table
+                    if (iBitNum >= 4 && (cc >> (iBitNum-4)) == 0xf) // first 4 bits are 1 - use long table
                     {
                         count = 16 - iBitNum;
                         codestart = cc << count;
@@ -1112,7 +1112,7 @@
                 while (iLen)
                 {
                     //               if (iBitNum > 6) // do long table
-                    if ((cc >> (iBitNum-5)) == 0x1f) // first 5 bits are 1 - use long table
+                    if (iBitNum >= 5 && (cc >> (iBitNum-5)) == 0x1f) // first 5 bits are 1 - use long table
                     {
                         count = iMaxLength - iBitNum;
                         codestart = cc << count;
@@ -1197,7 +1197,7 @@
                 iLen = *pBits++; // get number of codes for this bit length
                 while (iLen)
                 {
-                    if ((cc >> (iBitNum-6)) == 0x3f) // first 6 bits are 1 - use long table
+                    if (iBitNum >= 6 && (cc >> (iBitNum-6)) == 0x3f) // first 6 bits are 1 - use long table
                     {
                         count = 16 - iBitNum;
                         codestart = cc << count;
